contracts/PaymentTokens/EuroToken.sol

```diff
@@ -109,44 +126,57 @@ contract EuroToken is
     // Public functions
     ////////////////////////
 
-    /// @notice deposit 'amount' of EUR-T to address 'to'
+    /// @notice deposit 'amount' of EUR-T to address 'to', attaching correlating `reference` to LogDeposit event
     /// @dev deposit may happen only in case 'to' can receive transfer in token controller
     ///     by default KYC is required to receive deposits
-    function deposit(address to, uint256 amount)
+    function deposit(address to, uint256 amount, bytes32 reference)
         public
         only(ROLE_EURT_DEPOSIT_MANAGER)
         onlyIfDepositAllowed(to, amount)
+        acceptAgreement(to)
     {
         require(to != address(0));
         _balances[to] = add(_balances[to], amount);
         _totalSupply = add(_totalSupply, amount);
-        emit LogDeposit(to, msg.sender, amount);
+        emit LogDeposit(to, msg.sender, amount, reference);
         emit Transfer(address(0), to, amount);
     }
 
     /// @notice runs many deposits within one transaction
     /// @dev deposit may happen only in case 'to' can receive transfer in token controller
     ///     by default KYC is required to receive deposits
-    function depositMany(address[] to, uint256[] amount)
+    function depositMany(address[] to, uint256[] amount, bytes32[] reference)
         public
     {
         require(to.length == amount.length);
+        require(to.length == reference.length);
         for (uint256 i = 0; i < to.length; i++) {
-            deposit(to[i], amount[i]);
+            deposit(to[i], amount[i], reference[i]);
         }
     }
 
     /// @notice withdraws 'amount' of EUR-T by burning required amount and providing a proof of whithdrawal
-    /// @dev proof is provided in form of log entry. based on that proof backend will make a bank transfer
+    /// @dev proof is provided in form of log entry. based on that proof deposit manager will make a bank transfer
     ///     by default controller will check the following: KYC and existence of working bank account
     function withdraw(uint256 amount)
         public
         onlyIfWithdrawAllowed(msg.sender, amount)
+        acceptAgreement(msg.sender)
     {
         destroyTokensPrivate(msg.sender, amount);
         emit LogWithdrawal(msg.sender, amount);
     }
 
+    /// @notice issued by deposit manager when withdraw request was settled. typicaly amount that could be settled will be lower
+    ///         than amount withdrawn: banks charge negative interest rates and various fees that must be deduced
+    ///         reference number is attached that may be used to identify withdraw operation at deposit manager
+    function settleWithdraw(address from, uint256 amount, uint256 originalAmount, bytes32 withdrawTxHash, bytes32 reference)
+        public
+        only(ROLE_EURT_DEPOSIT_MANAGER)
+    {
+        emit LogWithdrawSettled(from, msg.sender, amount, originalAmount, withdrawTxHash, reference);
```

Comment:
This is also cool. We'll have to somehow import the bank csv for the withdrawals to write this event.

Reply:
this meant to be issued by withdraw service or backoffice app when transfer was sent... right now I think we should have a manual process?